### ISSUE TITLE
golang sql driver connect to clickhouse failed 

### ISSUE THREAD

**Issue by lightee2** [yandex/ClickHouse#3014]
hi,i am using golang sql driver to connect to clickhouse, an exception ocurred,the golang driver url :https://github.com/kshvakov/clickhouse. can anyone help me ?

beflow is the clickhouse server log. 
2018.09.01 08:17:32.252461 [ 20 ] <Trace> TCPHandlerFactory: TCP Request. Address: [::ffff:10.52.2.59]:63036
2018.09.01 08:18:17.761297 [ 20 ] <Debug> TCPHandler: Connected Golang SQLDriver version 1.1.0, revision: 54213, database: default, user: default.
2018.09.01 08:18:17.808393 [ 20 ] <Error> ServerErrorHandler: Code: 210, e.displayText() = DB::NetException: Connection reset by peer: while reading from socket ([::ffff:10.52.2.59]:63036), e.what() = DB::NetException, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x56d1436]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x22) [0x510bcd2]
2. /usr/bin/clickhouse-server(DB::ReadBufferFromPocoSocket::nextImpl()+0x248) [0x56ee2b8]
3. /usr/bin/clickhouse-server(DB::TCPHandler::runImpl()+0x57a) [0x2dc4dca]
4. /usr/bin/clickhouse-server(DB::TCPHandler::run()+0x2b) [0x2dc5acb]
5. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x58c56af]
6. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x16a) [0x58c5a8a]
7. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x77) [0x5966e97]
8. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x38) [0x5962d58]
9. /usr/bin/clickhouse-server() [0x938cc0f]
10. /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f0b1491e6db]
11. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f0b140a588f]

beflow is  the golang source code :
package main

import (
	"database/sql"
	"fmt"
	"log"

	"github.com/kshvakov/clickhouse"
)

func main() {
	connect, err := sql.Open("clickhouse", "tcp://192.168.0.88:9000?username=&compress=true&debug=true")
	if err != nil {
		log.Fatal(err)
	}
	if err := connect.Ping(); err != nil {
		if exception, ok := err.(*clickhouse.Exception); ok {
			fmt.Printf("[%d] %s \n%s\n", exception.Code, exception.Message, exception.StackTrace)
		} else {
			fmt.Println(err)
		}
		return
	}

}

**Comment by alexey-milovidov** [comment]
@kshvakov

**Comment by filimonov** [comment]
Todo section of [readme](https://github.com/kshvakov/clickhouse#todo) or go driver by @kshvakov. 
 can help you debug what's wrong.  

P.S. Remove `&compress=true` from connection string.

**Comment by filimonov** [comment]
I've check that:
```go
package main

import (
        "database/sql"
        "fmt"
        "log"
        "github.com/kshvakov/clickhouse"
)

func main() {
        connect, err := sql.Open("clickhouse", "tcp://127.0.0.1:9000?username=&compress=true&debug=true")
        if err != nil {
                log.Fatal(err)
        }
        if err := connect.Ping(); err != nil {
                if exception, ok := err.(*clickhouse.Exception); ok {
                        fmt.Printf("[%d] %s \n%s\n", exception.Code, exception.Message, exception.StackTrace)
                } else {
                        fmt.Println(err)
                }
                return
        }
}
```

And it perfectly works. Please provide the program output (you give only server logs) and check network connection from client to server using clickhouse-client.

**Comment by kshvakov** [comment]
@lightee2 Hi, are you connected to the container from the local host or is it remote server? 
What version of the ClickHouse server and Go client do you use? 
PS: Check the network connection from the client to server using clickhouse-client.

**Comment by lightee2** [comment]
@filimonov @kshvakov 
 the problem is so strange, i  run  the golang program on the server which is running clickhouse container successfully.,

**Comment by filimonov** [comment]
What platform and golang version do you use? For me it rather looks like something is wrong with golang installation or system inself because on the top of stacktrace there are 2 go core modules (garbage collector and bufio Writer). Normally core modules don't die in a such weird way.

**Comment by lightee2** [comment]
@filimonov thank you for your reply  :)
OS :windows 10 64bit
GO:go version go1.10.3 windows/386
docker + ku8s
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.7.4
 Git commit:   78d1802
 Built:        Tue Mar 14 09:47:15 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.7.4
 Git commit:   78d1802
 Built:        Tue Mar 14 09:47:15 2017
 OS/Arch:      linux/amd64

**Comment by filimonov** [comment]
Yep, i was afraid of something like that :D Windows host, multiple containers in docker, pretty old go version... And I bet containers communicate via windows-host (virtualbox /  Hyper-V) provided network... 

It's hard to imagine what exactly went bad there, but I can say that it doesn't look like [yandex/ClickHouse](https://github.com/yandex/ClickHouse) or [kshvakov/clickhouse](http://github.com/kshvakov/clickhouse) issue. 

God help you :)

**Comment by filimonov** [comment]
> i use telnet to test the remote server port 9000, the result is ok.

golang also connects properly and start communication properly, but dies after few packets. So telnet connection is bad test. You should try to connect using clickhouse-client from one container to another.

**Comment by lightee2** [comment]
@filimonov  @kshvakov 
thank you for you all reply,it is my turn to handle my own env problem.

**Comment by lightee2** [comment]
@filimonov the debug mode is ok . so odd
